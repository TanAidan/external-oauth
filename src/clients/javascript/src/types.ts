declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
