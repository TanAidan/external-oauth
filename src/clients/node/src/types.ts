declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
