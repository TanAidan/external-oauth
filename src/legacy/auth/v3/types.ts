import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
