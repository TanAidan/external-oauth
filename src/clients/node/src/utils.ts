import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
