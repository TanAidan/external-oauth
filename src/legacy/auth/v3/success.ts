import { Response } from 'express'
import { AuthSuccessRequest } from './types'
import { asyncMiddleware } from '../../../legacy/errorHandler'
import { updateAuth, TOAuthPayload } from '../clients/integrations'

export const authSuccess = asyncMiddleware(async (req: AuthSuccessRequest, res: Response) => {
  const { connectParams, setupId, authId, credentials, store, configuration } = req
  const buid = req.buid!

  const payload: TOAuthPayload = {
    connectParams,
    setupId,
    serviceName: buid,
    userId: authId,
    updatedAt: Date.now(),
    ...credentials
  }

  if (configuration && configuration.scopes) {
    console.log('[authSuccess] scopes', configuration.scopes)
    payload.scopes = configuration.scopes
  }

  if (req.tokenResponse) {
    payload.tokenResponseJSON = JSON.stringify(req.tokenResponse)
  }

  if (req.isCallback) {
    payload.callbackParamsJSON = JSON.stringify(req.query)
  }

  const params = {
    buid,
    authId,
    setupId,
    payload
  }

  await updateAuth({ ...params, store })

  res.header('Content-Type', 'text/html')
  res.render('auth/callback', { authId, error: '', error_description: '', integrationUuid: buid })
})
