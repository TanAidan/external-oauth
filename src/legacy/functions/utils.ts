// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
