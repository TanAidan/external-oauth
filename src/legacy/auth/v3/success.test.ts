import { MiddlewareTestHarness } from '../../../../tests/utils'
import { AuthSuccessRequest, EAuthType, TIntegrationConfig } from './types'
import { authSuccess } from './success'
// import { updateAuthV3 } from '../../clients/integrations'
import timekeeper from 'timekeeper'

jest.mock('../../clients/integrations')

describe('authSuccess', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  const setup = ({
    integrationConfig,
    tokenResponse,
    isCallback = false
  }: {
    integrationConfig?: TIntegrationConfig
    tokenResponse?: any
    isCallback?: boolean
  } = {}) =>
    new MiddlewareTestHarness({
      configureRequest: (req: AuthSuccessRequest) => {
        req.buid = 'test-buid'
        req.clientId = 'test-client-id'
        req.connectParams = {
          conParam: 'test-con-param'
        }
        req.setupId = 'test-setup-id'
        req.authId = 'test-auth-id'
        req.tokenResponse = tokenResponse
        req.isCallback = isCallback
        req.credentials = {
          accessToken: 'test-access-token',
          refreshToken: 'test-refresh-token',
          idToken: 'test-id-token',
          expiresIn: 1234
        }
        if (integrationConfig) {
          req.integrationConfig = integrationConfig
        }
      },
      testMiddleware: authSuccess
    })

  it('saves the credentials for the given authId', async () => {
    await timekeeper.withFreeze(new Date(12345678), async () =>
      setup()
        .get()
        .expect(200)
    )

    // expect(updateAuthV3).toMatchSnapshot()
  })

  it('saves the credentials with scopes for the given authId', async () => {
    const integrationConfig: TIntegrationConfig = {
      authType: EAuthType.OAuth2,
      config: {
        scope: ['users']
      }
    }
    await timekeeper.withFreeze(new Date(12345678), async () =>
      setup({ integrationConfig })
        .get()
        .expect(200)
    )

    // expect(updateAuthV3).toMatchSnapshot()
  })

  it('saves the credentials without scopes for the given authId', async () => {
    const integrationConfig: TIntegrationConfig = {
      authType: 'OAUTH2' as any,
      config: {}
    }
    await timekeeper.withFreeze(new Date(12345678), async () =>
      setup({ integrationConfig })
        .get()
        .expect(200)
    )

    // expect(updateAuthV3).toMatchSnapshot()
  })

  it('saves the credentials with token response', async () => {
    await timekeeper.withFreeze(new Date(12345678), async () =>
      setup({ tokenResponse: { body: { hey: 'there' } } })
        .get()
        .expect(200)
    )

    // expect(updateAuthV3).toMatchSnapshot()
  })

  it('renders the `callback` template with the authId and buid', async () => {
    const test = setup()

    await test
      .get()
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(200)

    expect(test.view).toMatchSnapshot()
  })

  describe('when the request is a callback request', () => {
    it('saves the query params as JSON along with the credentials', async () => {
      await timekeeper.withFreeze(new Date(12345678), async () =>
        setup({ isCallback: true })
          .get()
          .query({ some_param: 'param value' })
          .expect(200)
      )

      // expect(updateAuthV3).toMatchSnapshot()
    })
  })
})
