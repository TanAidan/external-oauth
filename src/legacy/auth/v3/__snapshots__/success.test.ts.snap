// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authSuccess renders the \`callback\` template with the authId and buid 1`] = `
Object {
  "name": "callback",
  "options": Object {
    "authId": "test-auth-id",
    "error": "",
    "error_description": "",
    "integrationUuid": "test-buid",
  },
}
`;

exports[`authSuccess saves the credentials for the given authId 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-auth-id",
        "buid": "test-buid",
        "clientId": "test-client-id",
        "servicesTableName": "services-table-name",
        "payload": Object {
          "accessToken": "test-access-token",
          "connectParams": Object {
            "conParam": "test-con-param",
          },
          "expiresIn": 1234,
          "idToken": "test-id-token",
          "refreshToken": "test-refresh-token",
          "serviceName": "test-buid",
          "setupId": "test-setup-id",
          "updatedAt": 12345678,
          "userId": "test-auth-id",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`authSuccess saves the credentials with scopes for the given authId 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-auth-id",
        "buid": "test-buid",
        "clientId": "test-client-id",
        "servicesTableName": "services-table-name",
        "payload": Object {
          "accessToken": "test-access-token",
          "connectParams": Object {
            "conParam": "test-con-param",
          },
          "expiresIn": 1234,
          "idToken": "test-id-token",
          "refreshToken": "test-refresh-token",
          "scopes": Array [
            "users",
          ],
          "serviceName": "test-buid",
          "setupId": "test-setup-id",
          "updatedAt": 12345678,
          "userId": "test-auth-id",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`authSuccess saves the credentials with token response 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-auth-id",
        "buid": "test-buid",
        "clientId": "test-client-id",
        "servicesTableName": "services-table-name",
        "payload": Object {
          "accessToken": "test-access-token",
          "connectParams": Object {
            "conParam": "test-con-param",
          },
          "expiresIn": 1234,
          "idToken": "test-id-token",
          "refreshToken": "test-refresh-token",
          "serviceName": "test-buid",
          "setupId": "test-setup-id",
          "tokenResponseJSON": "{\\"body\\":{\\"hey\\":\\"there\\"}}",
          "updatedAt": 12345678,
          "userId": "test-auth-id",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`authSuccess saves the credentials without scopes for the given authId 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-auth-id",
        "buid": "test-buid",
        "clientId": "test-client-id",
        "servicesTableName": "services-table-name",
        "payload": Object {
          "accessToken": "test-access-token",
          "connectParams": Object {
            "conParam": "test-con-param",
          },
          "expiresIn": 1234,
          "idToken": "test-id-token",
          "refreshToken": "test-refresh-token",
          "scopes": Array [],
          "serviceName": "test-buid",
          "setupId": "test-setup-id",
          "updatedAt": 12345678,
          "userId": "test-auth-id",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`authSuccess when the request is a callback request saves the query params as JSON along with the credentials 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authId": "test-auth-id",
        "buid": "test-buid",
        "clientId": "test-client-id",
        "servicesTableName": "services-table-name",
        "payload": Object {
          "accessToken": "test-access-token",
          "callbackParamsJSON": "{\\"some_param\\":\\"param value\\"}",
          "connectParams": Object {
            "conParam": "test-con-param",
          },
          "expiresIn": 1234,
          "idToken": "test-id-token",
          "refreshToken": "test-refresh-token",
          "serviceName": "test-buid",
          "setupId": "test-setup-id",
          "updatedAt": 12345678,
          "userId": "test-auth-id",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
